docs/concepts.md

```diff
@@ -0,0 +1,320 @@
+objects listening for events being scheduled or executed by other objects. The core
+of each simulated Entity became a large, hairy switch statement. The second problem
+was that FSMs and events could not easily represent that Replicas are created and
+destroyed during the life of the simulation. The assignment of Requests to Replicas
+is critical to provoking a realistic response from the Knative Pod Autoscaler.
+It is not enough to have an FSM representing "processing" as a state, it has to
+represent "processing on replica-4" as a state. But the existence and reachability
+of this state is contingent on the existence or non-existence of Replicas. This
+was easier to represent using Stocks and Movements that deal in Entities.
+
+Each Movement has four key values: Kind, OccursAt, From and To. The Kind is useful
+to group together Movements with different particulars. OccursAt is the point in
+simulation time at which the Movement is intended to occur. The "From" and "To"
+fields point to particular Stocks that the Environment will Remove() from and
+Add() to.
+
+Not included in Movement is any Entity. In the early "event" design for Skenario,
+the relationship was inverted so that Entities (called "Subjects" in an event)
+were responsible for all the logic of reacting to an event. This broke down because
+Entities would need to listen for events of interest, even if those events were
+not directly in their own scope of responsibility. For example, a Request would
+need to listen for the activation of Replicas, so that it could schedule itself
+to switch to the RequestProcessing state.
+
+Instead, as noted above, the responsibility for selecting which Entity to move
+rests with the Source Stock.
+
+#### Relationship to the Systems Dynamics concept of Flows
+
+In System Dynamics there's no "movement" concept and generally Entities are
+represented by a count or summation in a Stock. A Stock in System Dynamics is
+connected to other Stocks via Flows. Like Stocks these are functions of time
+elapsed, but they are derivatives, not integrals: rates of change, rather than
+accumulated amounts of change.
+
+This makes sense in a continuous setting, but does not make sense in a discrete
+setting. In this respect Movements can be thought of as discretized Flows, or
+alternatively as the inverse of a Flow. Where a Flow might expressed "X units
+per second", a Movement will instead look at the seconds until a particular
+Movement is to occur. When aggregated by time, Movements approximate a Flow.
+
+### Models
+
+Models are "the rest" of the code. Typically these own Stocks, wire dependencies
+and potentially maintain other state.
+
+In Skenario the main toplevel Models are the KnativeAutoscaler and the Cluster.
+These establish the various initial Stocks (eg. RequestsBuffered, ReplicasLaunching)
+that are used during the life of the simulation.
+
+Generally speaking, keep logic out of Models. Everything that can be expressed
```

Comment:
Surprising change of voice to imperative: "keep logic out of Models". "logic should be kept out of Models" reads better, I think.